# Overview
This script ranks evolutionary trees inferred from bulk DNA sequencing data by leveraging low-pass single-cell DNA sequencing (scDNA-seq) data. The method is designed to prioritize the most probable single-nucleotide variant (SNV) phylogenetic tree, helping to resolve ambiguities in tree inference. The script supports processing trees generated by **Conipher** and **Sapling** and ranks them based on likelihood of each tree given the scDNA-seq data.


## Table of Contents
- [Features](#features)
- [Methodology](#methodology)
- [Dependencies](#dependencies)
- [Installion](#installation)
- [I/O Data Formats](#io-data-formats)
    - [Input Data](#input-data)
        - [Conipher tree format](#conipher-tree-format)
        - [Sapling tree format](#sapling-tree-format)
        - [Read counts format](#read-counts-file-format)
    - [Output Data](#output-files)
- [Usage](#usage)
    - [Arborist CLI tool](#arborist-cli-tool)
    - [`arborist` package](#arborist-package)


## Features
- Processes read count data to compute likelihood scores for a given tree.
- Assigns cells to their most probable clones and computes the posterior probability of assignment to each clone in the tree.
- Computes the likelihood of each tree.
- Outputs ranked trees and cell assignments.
- Functions to compute the cell mutational burden (cmb) on the output of `arborist` for additional valdiation.


---

## Methodology

1. **Tree Parsing**  
   - Reads and stores parent-child relationships.

2. **Probability Calculation**  
   - Constructs an evolution matrix mapping clones to ancestors.
   - Assigns cells to clusters based on read count data.
   - Computes likelihood scores using a binomial probability model.

3. **Tree Ranking**  
   - Assigns each cell to the most probable clone.
   - Computes raw and normalized probabilities for trees.
   - Sorts trees based on probability scores.

4. **Entropy Calculation**  
   - Computes entropy for each cell assignment to quantify uncertainty.

---

## Dependencies
The script requires the following Python libraries:
- `python>=3.7`
- `pandas`
- `numpy`
- `scipy`
- `pygrahviz`



## Installation

Clone the repository from github.

```bash
git clone https://github.com/VanLoo-lab/Arborist.git
cd Arborist

```

Install package locally. Note: dependencie will be installed automatically via pip.

```bash
pip install .
```


## I/O Data Formats

### Input Data

#### **Read Counts File Format**
The input read counts file should be in CSV format or `panda.DataFrame` and contain the following columns:
| Column | Description |
|---------|------------|
| `cell` | Unique identifier for the cell to which . |
| `cluster` | Cluster ID to which the c belongs. |
| `total` | Total number of reads. |
| `alt` | Number of variant reads. |



#### **Conipher Tree Format**
- Trees are labeled numerically (e.g., `#tree 1`).
- Each tree consists of space-separated parent-child relationships.
- Trees are separated by blank lines.

**Example:**
```
#tree 1
7 2
6 7
2 0
2 4
7 1
7 3
```

#### **Sapling Tree Format**
- Trees start with a backbone tree label, followed by likelihood scores.
- The format includes space-separated parent-child relationships. 
- -1 represents the normal clone.

**Example:**
```
backbone tree 0, llh: -471.368062
-1 6
2 0
2 4
3 1
5 3
6 2
6 7
7 5
```

---

## Output Data
Example input and output files can be found in the `example/input` and `example/output` directories.
<!-- 
| File | Description |
|------|------------|
| `ranked_trees.csv` | Ranked trees with their probabilities. |
| `top20_trees_cell_assignment.csv` | Cell assignments for the top 20 ranked trees. |
| `Entropy.csv` | Entropy estimates for cell assignments. | -->


## Usage

### Arborist CLI tool
```
arborist -h                                                                                       usage: arborist [-h] --read_counts READ_COUNTS -T TREES [--sapling] [--alpha ALPHA] [--topn TOPN] [--ranking RANKING] [--cell_assign CELL_ASSIGN]

Tree ranking script

options:
  -h, --help            show this help message and exit
  --read_counts READ_COUNTS
                        Path to read counts CSV file with columns 'cell', 'cluster', 'total', 'alt'
  -T, --trees TREES     Path to tree file
  --sapling             Use sapling format for tree edges, otherwise conipher format is assumed.
  --alpha ALPHA         Per base sequencing error
  --topn TOPN           Filter only the top n trees, default is 25.
  --ranking RANKING     Path to where tree ranking output should be saved
  --cell_assign CELL_ASSIGN
                        Path to where cell assignments output should be saved
```

#### Example
```bash
   arborist -R {read_count_fname} -T {trees_fname} \
   --alpha 0.001 --ranking {output_tree_rank_fname} \
   --draw {output_png_or_dot_fname} \
   --cell-assign {output_cell_assignment_fname} --verbose
```

### CMB CLI tool usage
```
cmb -h                                                                                        
usage: cmb [-h] -R READ_COUNTS -T TREES [--cell-assign CELL_ASSIGN] [--sapling] [--min-cells MIN_CELLS] [-o OUT] [--conipher]

options:
  -h, --help            show this help message and exit
  -R READ_COUNTS, --read_counts READ_COUNTS
                        Path to read counts CSV file with columns 'cell','snv', 'cluster', 'total', 'alt'
  -T TREES, --trees TREES
                        Path to tree file
  --cell-assign CELL_ASSIGN
                        Path to where cell assignments, with columns 'cell', 'snv', 'tree', 'clone'
  --sapling             Use sapling format for tree edges, otherwise conipher format is assumed.
  --min-cells MIN_CELLS
                        minimum number of cells to compute the scores for a clade
  -o OUT, --out OUT     filename where cmb values will be saved
  --conipher            if the tree is in conipher format

```

#### Example
```bash
   cmb -R {read_count_fname} -T {trees_fname} \
   --cell-assign {output_cell_assignment_fname} \
   --min-cells 20 -o {output_cmb_fname}
```

### `arborist` Package

Run `arborist` on a set of candidate trees.
```python
import pandas as pd 
from arborist import read_tree_edges_conipher, rank_trees

trees = read_tree_edges_conipher("my_conipher_file.txt")
read_counts = pd.read_csv("my_read_counts.csv")
alpha= 0.001

tree_ranking, cell_assign = rank_trees(trees, read_counts, alpha, topn=5)

print(tree_ranking.head())

print(cell_assign.head())


```

Perform validation on the trees using the cell mutational burden `cmb` metric.
```python
from arborist.cmb import cmb

#read a mapping of cluster to a list of SNVs from the read_counts
psi = dat[["snv", "cluster"]].drop_duplicates()
psi = dict(zip(psi["snv"], psi["cluster"]))
clade_snvs = defaultdict(list)
for j, cluster in psi.items():
   clade_snvs[cluster].append(j)

#minumum number of cells within/outside of clade needed to compute the metric.
min_cells = 20

#compute the cmb of each tree, clade and cell in the arborist output dataframe cell_assign
cmb_df = cmb(cell_assign,trees,clade_snvs, read_counts, min_cells)

print(cmb_df)
```



<!-- ## Example Run
To run the script, provide the required and optional arguments as follows:

```sh
python script.py \
    --read_counts data/read_counts.csv \
    --sapling data/sapling_tree.txt \
    --sapling_backbone data/sapling_backbone.txt \
    --conipher data/conipher_tree.txt \
    --output results/ -->
