
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Documentation for the arborist package">
      
      
        <meta name="author" content="Leah L. Weber">
      
      
        <link rel="canonical" href="https://vanloo-lab.github.io/Arborist/api/arborist/">
      
      
        <link rel="prev" href="../../io/">
      
      
        <link rel="next" href="../read_trees/">
      
      
        
      
      
      <link rel="icon" href="../../img/icon.svg">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>arborist - Arborist</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="brown" data-md-color-accent="green">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#arborist.arborist.arborist" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Arborist" class="md-header__button md-logo" aria-label="Arborist" data-md-component="logo">
      
  <img src="../../img/icon.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Arborist
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              arborist
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/VanLoo-lab/Arborist" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    VanLoo-lab/Arborist
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Arborist" class="md-nav__button md-logo" aria-label="Arborist" data-md-component="logo">
      
  <img src="../../img/icon.svg" alt="logo">

    </a>
    Arborist
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/VanLoo-lab/Arborist" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    VanLoo-lab/Arborist
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    CLI Tool
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../package-tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Package Tutorial
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    I/O Formats
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    API
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    API
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    arborist
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    arborist
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#arborist.arborist.arborist" class="md-nav__link">
    <span class="md-ellipsis">
      
        arborist
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../read_trees/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    read_trees
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../treefit/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TreeFit
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#arborist.arborist.arborist" class="md-nav__link">
    <span class="md-ellipsis">
      
        arborist
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>arborist</h1>

<div class="doc doc-object doc-function">



<a id="arborist.arborist.arborist"></a>
    <div class="doc doc-contents first">

        <p>Rank candidate SNV phylogenetic trees using scDNA-seq read counts.</p>
<p>This function evaluates a list of clonal trees using a variational inference
scheme and returns an evidence lower bound (ELBO)-based score for each tree.
An initial SNV-to-cluster prior is set given an initial SNV clustering, and both cell-to-clone and
SNV-to-cluster assignments are optimized under the model.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>tree_list</code>
            </td>
            <td>
                  <code>list of list of tuple</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Candidate phylogenetic trees to be ranked. Each tree is represented as
a list of directed edges <code>(parent, child)</code>. All trees must contain
the same set of clone identifiers.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>read_counts</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Data frame with columns <code>["snv", "cell", "alt", "total"]</code> in any order, 
containing per-cell read counts for each SNV.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>snv_clusters</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Data frame with columns <code>["snv", "cluster"]</code> in any order, giving an
initial hard assignment of SNVs to clusters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>alpha</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Per-base sequencing error rate used to compute log-likelihoods for
presence/absence of an SNV (default is <code>0.001</code>).</p>
              </div>
            </td>
            <td>
                  <code>0.001</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>max_iter</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of coordinate-ascent iterations in the variational
inference procedure (default is <code>10</code>).</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tolerance</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Convergence threshold on the change in ELBO between iterations
(default is <code>1.0</code>).</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gamma</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Prior probability mass placed on the initial SNV cluster assignment
in <code>q_y</code> (default is <code>0.7</code>). The remaining mass is spread
uniformly over alternative clusters.</p>
              </div>
            </td>
            <td>
                  <code>0.7</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>add_normal</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>True</code>, prepend a normal clone as a new root node to each tree in
<code>tree_list</code> (default is <code>False</code>).</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threads</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of threads to use for numba-parallelized computations
(default is <code>10</code>).</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>verbose</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>True</code>, enable informative logging messages during fitting
(default is <code>False</code>).</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>likelihoods</code></td>            <td>
                  <code><span title="dict">dict</span>[<span title="int">int</span>, <span title="float">float</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping from tree index to its ELBO (expected log joint) under the
variational posterior.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>best_fit</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code&gt;arborist.treefit.TreeFit&lt;/code&gt;" href="../treefit/#arborist.treefit.TreeFit">TreeFit</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><code>TreeFit</code> object for the top-ranked tree, containing the tree,
ELBO, posterior cell-to-clone distribution <code>q_z</code>, posterior
SNV-to-cluster distribution <code>q_y</code>, and associated index mappings.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>all_tree_fits</code></td>            <td>
                  <code><span title="dict">dict</span>[<span title="int">int</span>, <a class="autorefs autorefs-internal" title="&lt;code&gt;arborist.treefit.TreeFit&lt;/code&gt;" href="../treefit/#arborist.treefit.TreeFit">TreeFit</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping each tree index to its corresponding <code>TreeFit</code>
object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the candidate trees in <code>tree_list</code> do not all share the same
set of clone identifiers.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Notes</summary>
  <p>Only SNVs with at least one variant read across all cells are retained
for inference. SNVs with zero total variant counts are dropped from
<code>read_counts</code> before fitting.</p>
</details>

<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">likelihoods</span><span class="p">,</span> <span class="n">best_fit</span><span class="p">,</span> <span class="n">all_fits</span> <span class="o">=</span> <span class="n">arborist</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">tree_list</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">read_counts</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">snv_clusters</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_iter</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">gamma</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</code></pre></div>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/arborist/arborist.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">arborist</span><span class="p">(</span>
    <span class="n">tree_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
    <span class="n">read_counts</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">snv_clusters</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">alpha</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
    <span class="n">max_iter</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">tolerance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">gamma</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
    <span class="n">add_normal</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">threads</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Rank candidate SNV phylogenetic trees using scDNA-seq read counts.</span>

<span class="sd">    This function evaluates a list of clonal trees using a variational inference</span>
<span class="sd">    scheme and returns an evidence lower bound (ELBO)-based score for each tree.</span>
<span class="sd">    An initial SNV-to-cluster prior is set given an initial SNV clustering, and both cell-to-clone and</span>
<span class="sd">    SNV-to-cluster assignments are optimized under the model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    tree_list : list of list of tuple</span>
<span class="sd">        Candidate phylogenetic trees to be ranked. Each tree is represented as</span>
<span class="sd">        a list of directed edges ``(parent, child)``. All trees must contain</span>
<span class="sd">        the same set of clone identifiers.</span>
<span class="sd">    read_counts : pandas.DataFrame</span>
<span class="sd">        Data frame with columns ``[&quot;snv&quot;, &quot;cell&quot;, &quot;alt&quot;, &quot;total&quot;]`` in any order, </span>
<span class="sd">        containing per-cell read counts for each SNV.</span>
<span class="sd">    snv_clusters : pandas.DataFrame</span>
<span class="sd">        Data frame with columns ``[&quot;snv&quot;, &quot;cluster&quot;]`` in any order, giving an</span>
<span class="sd">        initial hard assignment of SNVs to clusters.</span>
<span class="sd">    alpha : float, optional</span>
<span class="sd">        Per-base sequencing error rate used to compute log-likelihoods for</span>
<span class="sd">        presence/absence of an SNV (default is ``0.001``).</span>
<span class="sd">    max_iter : int, optional</span>
<span class="sd">        Maximum number of coordinate-ascent iterations in the variational</span>
<span class="sd">        inference procedure (default is ``10``).</span>
<span class="sd">    tolerance : float, optional</span>
<span class="sd">        Convergence threshold on the change in ELBO between iterations</span>
<span class="sd">        (default is ``1.0``).</span>
<span class="sd">    gamma : float, optional</span>
<span class="sd">        Prior probability mass placed on the initial SNV cluster assignment</span>
<span class="sd">        in ``q_y`` (default is ``0.7``). The remaining mass is spread</span>
<span class="sd">        uniformly over alternative clusters.</span>
<span class="sd">    add_normal : bool, optional</span>
<span class="sd">        If ``True``, prepend a normal clone as a new root node to each tree in</span>
<span class="sd">        ``tree_list`` (default is ``False``).</span>
<span class="sd">    threads : int, optional</span>
<span class="sd">        Number of threads to use for numba-parallelized computations</span>
<span class="sd">        (default is ``10``).</span>
<span class="sd">    verbose : bool, optional</span>
<span class="sd">        If ``True``, enable informative logging messages during fitting</span>
<span class="sd">        (default is ``False``).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    likelihoods : dict[int, float]</span>
<span class="sd">        Mapping from tree index to its ELBO (expected log joint) under the</span>
<span class="sd">        variational posterior.</span>
<span class="sd">    best_fit : TreeFit</span>
<span class="sd">        ``TreeFit`` object for the top-ranked tree, containing the tree,</span>
<span class="sd">        ELBO, posterior cell-to-clone distribution ``q_z``, posterior</span>
<span class="sd">        SNV-to-cluster distribution ``q_y``, and associated index mappings.</span>
<span class="sd">    all_tree_fits : dict[int, TreeFit]</span>
<span class="sd">        Dictionary mapping each tree index to its corresponding ``TreeFit``</span>
<span class="sd">        object.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the candidate trees in ``tree_list`` do not all share the same</span>
<span class="sd">        set of clone identifiers.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Only SNVs with at least one variant read across all cells are retained</span>
<span class="sd">    for inference. SNVs with zero total variant counts are dropped from</span>
<span class="sd">    ``read_counts`` before fitting.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; likelihoods, best_fit, all_fits = arborist(</span>
<span class="sd">    ...     tree_list,</span>
<span class="sd">    ...     read_counts,</span>
<span class="sd">    ...     snv_clusters,</span>
<span class="sd">    ...     alpha=0.001,</span>
<span class="sd">    ...     max_iter=25,</span>
<span class="sd">    ...     gamma=0.7,</span>
<span class="sd">    ... )</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>

    <span class="n">numba</span><span class="o">.</span><span class="n">set_num_threads</span><span class="p">(</span><span class="n">threads</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">add_normal</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Appending normal clone to candidate trees...&quot;</span><span class="p">)</span>
        <span class="n">tree_list</span> <span class="o">=</span> <span class="n">add_normal_clone</span><span class="p">(</span><span class="n">tree_list</span><span class="p">)</span>

    <span class="n">tree</span> <span class="o">=</span> <span class="n">tree_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># assume root is normal</span>
    <span class="n">temp_tree</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
    <span class="n">normal</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">temp_tree</span> <span class="k">if</span> <span class="n">temp_tree</span><span class="o">.</span><span class="n">in_degree</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">clone_set</span> <span class="o">=</span> <span class="n">tree_to_clone_set</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">tree_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">tree_to_clone_set</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span> <span class="o">!=</span> <span class="n">clone_set</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;All trees must have the same set of clones.&quot;</span><span class="p">)</span>

    <span class="n">clones</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">clone_set</span><span class="p">)</span>

    <span class="n">clusters</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">clones</span> <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">normal</span><span class="p">]</span>

    <span class="c1"># Filter read_counts to only include cells and SNVs present in the tree</span>

    <span class="n">clones</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">clusters</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">clone_to_idx</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">clones</span><span class="p">)}</span>
    <span class="n">cluster_to_idx</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">clusters</span><span class="p">)}</span>


    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Removing SNVs with 0 variant reads across all cells...&quot;</span><span class="p">)</span>

    <span class="n">alt_sum</span> <span class="o">=</span> <span class="n">read_counts</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;snv&quot;</span><span class="p">)[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">valid_snvs</span> <span class="o">=</span> <span class="n">alt_sum</span><span class="p">[</span><span class="n">alt_sum</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>


    <span class="c1"># valid SNVs are SNVs that have at least one variant read across all cells</span>
    <span class="c1"># otherwise we have no signal to place them in the tree</span>

    <span class="n">read_counts</span> <span class="o">=</span> <span class="n">read_counts</span><span class="p">[</span><span class="n">read_counts</span><span class="p">[</span><span class="s2">&quot;snv&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">valid_snvs</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># appends columns log_absent and log_present to read_counts</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Caching log-likelihoods for presence/absence...&quot;</span><span class="p">)</span>
    <span class="n">read_counts</span><span class="p">,</span> <span class="n">cell_to_idx</span><span class="p">,</span> <span class="n">snv_to_idx</span> <span class="o">=</span> <span class="n">precompute_log_likelihoods</span><span class="p">(</span>
        <span class="n">read_counts</span><span class="p">,</span> <span class="n">alpha</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Initializing the SNV cluster assignment prior...&quot;</span><span class="p">)</span>
    <span class="n">q_y_init</span> <span class="o">=</span> <span class="n">initialize_q_y</span><span class="p">(</span><span class="n">snv_clusters</span><span class="p">,</span> <span class="n">cluster_to_idx</span><span class="p">,</span> <span class="n">snv_to_idx</span><span class="p">,</span> <span class="n">gamma</span><span class="p">)</span>
    <span class="n">cell_idx</span><span class="p">,</span> <span class="n">snv_idx</span><span class="p">,</span> <span class="n">log_like_matrix</span> <span class="o">=</span> <span class="n">build_sparse_input</span><span class="p">(</span>
        <span class="n">read_counts</span><span class="p">,</span> <span class="n">cell_to_idx</span><span class="p">,</span> <span class="n">snv_to_idx</span>
    <span class="p">)</span>

    <span class="n">n_cells</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_to_idx</span><span class="p">)</span>
    <span class="n">n_snvs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">snv_to_idx</span><span class="p">)</span>
    <span class="n">n_clones</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">clones</span><span class="p">)</span>


    <span class="n">cell_ptr</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">snv_index_cell_sort</span><span class="p">,</span> <span class="n">log_like_matrix_cell_sort</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">build_index_pointers</span><span class="p">(</span>
            <span class="n">cell_idx</span><span class="p">,</span> <span class="n">snv_idx</span><span class="p">,</span> <span class="n">n_cells</span><span class="p">,</span> <span class="n">log_like_matrix</span><span class="o">=</span><span class="n">log_like_matrix</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">snv_ptr</span><span class="p">,</span> <span class="n">_</span> <span class="p">,</span> <span class="n">cell_index_snv_sort</span><span class="p">,</span> <span class="n">log_like_matrix_snv_sort</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">build_index_pointers</span><span class="p">(</span><span class="n">snv_idx</span><span class="p">,</span> <span class="n">cell_idx</span><span class="p">,</span> <span class="n">n_snvs</span><span class="p">,</span> <span class="n">log_like_matrix</span><span class="o">=</span><span class="n">log_like_matrix</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">best_likelihood</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span>
    <span class="n">likelihoods</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">all_tree_fits</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">run</span> <span class="o">=</span> <span class="n">run_variational_inference</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Starting Arborist...&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of candidate trees: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of clones: </span><span class="si">{</span><span class="n">n_clones</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of SNV clusters: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">cluster_to_idx</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of cells: </span><span class="si">{</span><span class="n">n_cells</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of SNVs: </span><span class="si">{</span><span class="n">n_snvs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># else:</span>
    <span class="c1">#     print(&quot;Running Arborist in cell MAP assignment mode....&quot;)</span>
    <span class="c1">#     run = run_simple_max_likelihood</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tree_list</span><span class="p">):</span>

        <span class="n">presence</span> <span class="o">=</span> <span class="n">enumerate_presence</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">clone_to_idx</span><span class="p">,</span> <span class="n">cluster_to_idx</span><span class="p">)</span>

        <span class="n">expected_log_like</span><span class="p">,</span> <span class="n">q_z</span><span class="p">,</span> <span class="n">q_y</span> <span class="o">=</span> <span class="n">run</span><span class="p">(</span>
            <span class="n">presence</span><span class="p">,</span>
            <span class="n">log_like_matrix_cell_sort</span><span class="p">,</span>
            <span class="n">log_like_matrix_snv_sort</span><span class="p">,</span>
            <span class="n">cell_idx</span><span class="o">=</span><span class="n">cell_index_snv_sort</span><span class="p">,</span>
            <span class="n">snv_idx</span><span class="o">=</span><span class="n">snv_index_cell_sort</span><span class="p">,</span>
            <span class="n">n_cells</span><span class="o">=</span><span class="n">n_cells</span><span class="p">,</span>
            <span class="n">n_snvs</span><span class="o">=</span><span class="n">n_snvs</span><span class="p">,</span>
            <span class="n">n_clones</span><span class="o">=</span><span class="n">n_clones</span><span class="p">,</span>
            <span class="n">q_y_init</span><span class="o">=</span><span class="n">q_y_init</span><span class="p">,</span>
            <span class="n">cell_ptr</span><span class="o">=</span><span class="n">cell_ptr</span><span class="p">,</span>
            <span class="n">snv_ptr</span><span class="o">=</span><span class="n">snv_ptr</span><span class="p">,</span>
            <span class="n">max_iter</span><span class="o">=</span><span class="n">max_iter</span><span class="p">,</span>
            <span class="n">tolerance</span><span class="o">=</span><span class="n">tolerance</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">tfit</span> <span class="o">=</span> <span class="n">TreeFit</span><span class="p">(</span>
            <span class="n">tree_list</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span>
            <span class="n">idx</span><span class="p">,</span>
            <span class="n">expected_log_like</span><span class="p">,</span>
            <span class="n">q_z</span><span class="p">,</span>
            <span class="n">q_y</span><span class="p">,</span>
            <span class="n">cell_to_idx</span><span class="p">,</span>
            <span class="n">snv_to_idx</span><span class="p">,</span>
            <span class="n">clone_to_idx</span><span class="p">,</span>
            <span class="n">cluster_to_idx</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">all_tree_fits</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">tfit</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tree </span><span class="si">{</span><span class="n">idx</span><span class="si">}</span><span class="s2"> fit wtih ELBO: </span><span class="si">{</span><span class="n">expected_log_like</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">likelihoods</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">expected_log_like</span>
        <span class="k">if</span> <span class="n">expected_log_like</span> <span class="o">&gt;</span> <span class="n">best_likelihood</span><span class="p">:</span>
            <span class="n">best_fit</span> <span class="o">=</span> <span class="n">tfit</span>
            <span class="n">best_likelihood</span> <span class="o">=</span> <span class="n">expected_log_like</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Done fitting all </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">tree_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> candidate trees!&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">likelihoods</span><span class="p">,</span> <span class="n">best_fit</span><span class="p">,</span> <span class="n">all_tree_fits</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["content.tooltips", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest"], "search": "../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>